packages/opencensus-instrumentation-http2/src/http2.ts

```diff
@@ -0,0 +1,283 @@
+        }
+
+        const request = original.apply(this, arguments);
+        plugin.tracer.wrapEmitter(request);
+
+        const traceOptions = {
+          name: `${headers[':method'] ? headers[':method'] : 'GET'} ${
+              headers[':path']}`,
+          type: 'CLIENT',
+        };
+
+        // Checks if this outgoing request is part of an operation by checking
+        // if there is a current root span, if so, we create a child span. In
+        // case there is no root span, this means that the outgoing request is
+        // the first operation, therefore we create a root span.
+        if (!plugin.tracer.currentRootSpan) {
+          return plugin.tracer.startRootSpan(
+              traceOptions,
+              plugin.makeHttp2RequestTrace(request, headers, plugin));
+        } else {
+          const span = plugin.tracer.startChildSpan(
+              traceOptions.name, traceOptions.type);
+          return (plugin.makeHttp2RequestTrace(request, headers, plugin))(span);
+        }
+      };
+    };
+  }
+
+  makeHttp2RequestTrace(
+      request: http2.ClientHttp2Stream, headers: http2.IncomingHttpHeaders,
+      plugin: Http2Plugin): types.Func<http2.ClientHttp2Stream> {
+    return (span: types.Span): http2.ClientHttp2Stream => {
+      if (!span) return request;
+
+      const setter = {
+        setHeader(name: string, value: string) {
+          headers[name] = value;
+        }
+      } as types.HeaderSetter;
+
+      const propagation = plugin.tracer.propagation;
+      if (propagation) {
+        propagation.inject(setter, span.spanContext);
+      }
+
+      request.on('response', (responseHeaders: http2.IncomingHttpHeaders) => {
+        span.addAttribute(
+            Http2Plugin.ATTRIBUTE_HTTP_STATUS_CODE,
+            `${responseHeaders[':status']}`);
+        span.status = plugin.traceStatus(+responseHeaders[':status']);
+      });
+
+      request.on('end', () => {
+        const userAgent =
+            (headers['user-agent'] || headers['User-Agent'] || null) as string;
+
+        span.addAttribute(
+            Http2Plugin.ATTRIBUTE_HTTP_HOST, plugin.clientUrl.host);
+        span.addAttribute(
+            Http2Plugin.ATTRIBUTE_HTTP_METHOD, `${headers[':method']}`);
+        span.addAttribute(
+            Http2Plugin.ATTRIBUTE_HTTP_PATH, `${headers[':path']}`);
+        span.addAttribute(
+            Http2Plugin.ATTRIBUTE_HTTP_ROUTE, `${headers[':path']}`);
+        span.addAttribute(Http2Plugin.ATTRIBUTE_HTTP_USER_AGENT, userAgent);
+
+        span.addMessageEvent(
+            'MessageEventTypeSent', uuid.v4().split('-').join(''));
+
+        span.end();
+      });
+
+      request.on('error', (err: Error) => {
+        span.addAttribute(HttpPlugin.ATTRIBUTE_HTTP_ERROR_NAME, err.name);
+        span.addAttribute(HttpPlugin.ATTRIBUTE_HTTP_ERROR_MESSAGE, err.message);
+
+        span.end();
+      });
+
+      return request;
+    };
+  }
+
+  patchCreateServer() {
```

Comment:
All `patch` functions should be `private`.

Reply:
Ok, just changed that. As we needed `patchRequest()` and `patchEmit()` in the tests, I created a method `patchForTest()`that does what we need. It is intended for testing purposes only.